### PR TITLE
Categorical cross entropy loss

### PR DESCRIPTION
**Pull request by ndemashov**
This PR adds support of categorical cross entropy loss

### CODE REVIEW

**Comment by ndemashov** [comment]
Hi @Ubospica and @SiriusNEO!
Please review my categorical cross entropy loss implementation.

**Comment by SiriusNEO** [comment]
I notice that your implementation is: first convert one-hot to label, then use nll loss. But if we already have one-hot representation, would it be good to just use a negative hadamard product? Like
```
logits = bb.emit(log_softmax(predictions))
loss = bb.emit(-logits * targets)
```
Although maybe we need to deal with `ignore_index` and `reduction` in this case.

**Comment by SiriusNEO** [comment]
@ndemashov Recently for synchronizing with apache/tvm unity branch, we drop some commits in this fork, including the loss functions part. The backup can be found in this branch: https://github.com/mlc-ai/relax/tree/mlc-merge-drop-backup-04-13.
We will soon bring it back and at that time you can rebase this PR onto it. Sorry for the inconvenience!

**Comment by ndemashov** [comment]
@SiriusNEO I implemented the categorical cross-entropy loss function in the way you suggested. Thank you for your advice. Using the _with_reduction function, I was able to implement **redaction**. But I ran into the problem of implementing an **ignore_index**. I did not find a suitable method or function to null the target index corresponding to this parameter. Could you please help me or maybe I should go back to the first implementation?

**Comment by SiriusNEO** [comment]
@ndemashov Yes you're right, `ignore_index` is indeed a problem. In the [torch document](https://pytorch.org/docs/stable/generated/torch.nn.CrossEntropyLoss.html), the `ignore_index` is invalid when `target` is class probabilities (i.e. categorical cross entropy loss); And in the [keras document](https://www.tensorflow.org/api_docs/python/tf/keras/losses/CategoricalCrossentropy) they don't provide `ignore_index`.

So IMO, we can **remove this argument**. And also if you want to reserve it, we can fall back to your first implementation if `ignore_index` is in a valid range:
```
if self.ignore_index >= 0:
   (your first implementation)
else:
   (your second implementation)
```
The main concern for the first implementation is the efficiency. But I think it's not a big problem. You can make your decision.

**Comment by ndemashov** [comment]
@SiriusNEO It is done. Please review the implementation and let me know how I can rebase my PR. Thank you for the help.

**Comment by ndemashov** [comment]
@tqchen @Ubospica @SiriusNEO Could you please clarify me when I should rebase onto `mlc-ai/mlc` branch? Probably should I create a new PR to `unity` branch in `apache/tvm` repository?

**Comment by SiriusNEO** [comment]
@ndemashov Sorry for the late reply. Please PR to the unity branch directly since https://github.com/apache/tvm/pull/14668 is merged. Mlc doesn't have loss function now but we will rebase onto unity in the future.

**Comment by SiriusNEO** [comment]
And this PR can be closed if there is no further problem
